beps/lib-log/1376_sensitive_data_exclusion.md

```diff
@@ -0,0 +1,198 @@
+public type Replacement record {|
+    string|ReplacementFunction replacement;
+|};
+
+# Masking strategy for sensitive data
+public type MaskingStrategy EXCLUDE|Replacement;
+
+# Represents sensitive data with a masking strategy
+#
+# + strategy - The masking strategy to apply (default: EXCLUDE)
+public type SensitiveDataConfig record {|
+    MaskingStrategy strategy = EXCLUDE;
+|};
+
+# Marks a record field or type as sensitive, excluding it from log output
+#
+# + strategy - The masking strategy to apply (default: EXCLUDE)
+public const annotation SensitiveDataConfig SensitiveData on record field, type;
+```
+
+> **Note**: The `SensitiveData` annotation can be applied to both record fields and types. Field level annotations take precedence over type-level annotations, allowing for granular control over sensitive data exclusion.
+
+#### Sensitive Data Exclusion Strategies
+
+1. **EXCLUDE**: Completely removes the sensitive data from the log output. This is the default behaviour.
+2. **REPLACE**: Replacement can be a string or a function:
+   - If a string is provided, it replaces the entire sensitive data with that string.
+   - If a function is provided, the original value is passed to the function, and the returned value is used in the masked value.
+
+#### Usage Examples
+
+```ballerina
+@log:SensitiveData {
+    strategy: {
+        replacement: "***********"
+    }
+}
+public type ApiKey string;
+
+@log:SensitiveData
+public type Password string;
+
+isolated function maskEmail(string email) returns string {
+    int? atIndex = email.indexOf("@");
+    if atIndex is int {
+        return email.substring(0, 1) + "****" + email.substring(atIndex);
+    }
+    return email;
+}
+
+type UserCredentials record {
+    string username;
+    Password password;
+    ApiKey apiKey;
+    @log:SensitiveData {
+        strategy: {
+            replacement: maskEmail
+        }
+    }
+    string email;
+};
+
+public function main() {
+    Password userPassword = "mySecretPassword123";
+    ApiKey key = "sk_live_abc123xyz789";
+    UserCredentials user = {
+        username: "john_doe",
+        password: "secret123",
+        apiKey: "sk_live_abc123xyz",
+        email: "john@example.com"
+    };
+
+    log:printInfo("User login attempt", password = userPassword, apiKey = key);
+    // Output: time=2025-08-06T10:30:00.000Z level=INFO message="User login attempt" apiKey="[REDACTED]"
+
+    log:printInfo("User authentication", user = user);
+    // Output: time=2025-08-06T10:30:00.000Z level=INFO message="User authentication"
+    //         user = {"username": "john_doe", "apiKey": "[REDACTED]", "email": "j****@gmail.com"}
+}
+```
+
+#### Configuration to Enable/Disable Sensitive Data Exclusion
+
+By default, sensitive data exclusion is enabled. However, users can configure the logging package to disable this feature if needed. The `ballerina/log` package provide a global configuration option to control sensitive data exclusion behavior.
```

Comment:
What is the usecase for users to enable sensitive data logging ?

Reply:
The alternative section covers this where we can pass the masking function to the custom logger. But that is not intutive enough. Since Ballerina is type oriented, following approach seems to be suitable for me.

Reply:
I think having a configuration that enables sensible data masking and make it disabled by default is the right way to go. MI has the same experience. https://mi.docs.wso2.com/en/latest/administer/logging-and-monitoring/logging/masking-sensitive-information-in-logs/